### PR TITLE
feat: narrow ResultInterface to concrete implementation after calling `isSucceeded()` and `isFailed()`

### PR DESCRIPTION
**Pull request by simPod**

### CODE REVIEW

File: tests/static-analysis/Result/result_interface.php [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+use Psl\Result\Failure;
+use Psl\Result\ResultInterface;
+use Psl\Result\Success;
+
+/**
+ * @param ResultInterface<int> $result
+ *
+ * @throws Exception
+ *
+ * @return Failure<int, Exception>
```
Comment (veewee) [inline tests/static-analysis/Result/result_interface.php:17]: Stange : Shouldn't psalm complain about the second generic parameter `Exception` that does not exist here?

**Review by veewee — COMMENTED** [review]
Thanks for the PR. 
Seems like a nice addition!

**Review by veewee — APPROVED** [review]

**Comment by azjezz** [comment]
I think we need to revert this 😅 `ResultInterface` is part of the public API, and while it is quite unlikey, someone could re-implement `Success`/`Failure`, e.g `MySuccess`/`MyFailure`, in which case, `MySuccess::isSucceeded` would return true, but that does not mean its an instance of `Success` ( that is what psalm would think ).

repro:

```
function foo(ResultInterface $result): Success {
  if ($result->isSucceeded()) {
    return $result;
  }

  exit('no');
}

foo($mySuccessImplementation); // throws an exception
```

[*CRIES*](https://wiki.php.net/rfc/sealed_classes)

**Comment by simPod** [comment]
Hm, that's true. 😢<sup>2</sup>

Should I send revert PR?

**Comment by veewee** [comment]
Woops, merged a bit too fast - sorry.
A revert PR is welcome. 😢 **3

**Comment by simPod** [comment]
There u go https://github.com/azjezz/psl/pull/469

**Comment by simPod** [comment]
BTW is there any chance the RFC is accepted in the future? (like it was missing something that could be fixed)

**Pull request merged**